### PR TITLE
Added a data set builder for the classifier.

### PR DESCRIPTION
**Pull request by ingridel**
Added a data set builder for the classifier and the corresponding testing module.

### CODE REVIEW

File: betel/classifier_data_set_builder.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from betel import data_set_builder
+
+
+class ClassifierDataSetBuilder(data_set_builder.DataSetBuilder):
+    """A class for splitting classifier data into train-validation-test sets."""
+
+    def __init__(self, input_dir: pathlib.Path, storage_dir: pathlib.Path, classes: [str] = None):
+        """Constructor.
+
+        :param input_dir: data to be split (output of the scraper)
```
Comment (tpetmanson) [inline betel/classifier_data_set_builder.py]: I think it is not super relevant for Github projects, but the Google's style for docstrings is different: go/pystyle#function-docs

File: betel/data_set_builder.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        self._input_file = input_dir / utils.SCRAPER_INFO_FILE_NAME
+
+    def split(self, data: pd.DataFrame) -> Dict[str, pd.DataFrame]:
+        """Randomly splits a list of data into 3
+        (train-validation-test).
+
+        :param data: the list to be split
+        :return: the train-validation-test split in Dict format
+        """
+        app_list = self._sort(data)
```
Comment (tpetmanson) [inline betel/data_set_builder.py:41]: It looks that on the next line you randomly shuffle the dataset? In that case the sorting can be omitted.

File: betel/data_set_builder.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        app_list = app_list.sample(frac=1, random_state=self._RANDOM_SEED)
+
+        index_1 = int(0.7 * len(app_list))
+        index_2 = int(0.85 * len(app_list))
+        train = app_list.iloc[:index_1]
```
Comment (tpetmanson) [inline betel/data_set_builder.py]: Just a thought, maybe any of sklearns train-test splitting functions can be used here, if that makes the code simpler? They also have the option for stratified sampling to get simlar representation of labels in train-test-validation. However, these functions split only into train-test so they would need to applied twice.

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#sklearn.model_selection.train_test_split

Just a thought, it's your call whether to use them.

File: betel/data_set_builder.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import pathlib
+import abc
+from typing import Dict
+import pandas as pd
+from betel import utils
+from betel import info_files_helpers
+
+
+class DataSetBuilder(metaclass=abc.ABCMeta):
```
Comment (tpetmanson) [inline betel/data_set_builder.py:10]: In case ClassifierDataSetBuilder would be the only class to subclass DataSetBuilder, it might be shorter/easier to to merge them into one class. No need to rewrite it though if you prefer this way.

File: test/classifier_data_set_builder_test.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        _create_icons(APP_LIST, input_dir)
+        expected_icons = _get_expected_icons(storage_dir, classes)
+
+        classifier_builder = classifier_data_set_builder.ClassifierDataSetBuilder(
+            input_dir, storage_dir, classes=classes
+        )
+
+        classifier_builder.split_and_build_data_sets()
+
+        assert all(icon.exists() for icon in expected_icons)
```
Comment (tpetmanson) [inline test/classifier_data_set_builder_test.py]: It is not super clear to me what exactly expected_icons are as you construct this list programmatically. If it is feasible, I would recommend giving them verbatim so it would be easier to see what is expected just by looking at the code. I'd recommend the same for the other tests.

**Review by tpetmanson — APPROVED** [review]
Code looks good overall! I left a few suggestions and random thoughts. Use your own judgement whether to address them or not.

File: betel/classifier_data_set_builder.py (hunk 1 above)
Comment (ingridel) [reply betel/classifier_data_set_builder.py]: Unfortunately, I do not have access to that. I used [this styleguide](https://google.github.io/styleguide/pyguide.html#typing-imports) and some suggestions from Seb&Momo.

File: betel/data_set_builder.py (hunk 2 above)
Comment (ingridel) [reply betel/data_set_builder.py:41]: Sorting in the beginning helps with the case when I get the same data set, but with ids in different order to generate the same split every time. As for the shuffling, it is always done with the same seed, so the property of a consistent split is still maintained.

File: betel/data_set_builder.py (hunk 4 above)
Comment (ingridel) [reply betel/data_set_builder.py:10]: I was thinking of another DataSetBuilder subclass for the image similarity model after the classifier, this is why I defined a parent class.

File: test/classifier_data_set_builder_test.py (hunk 5 above)
Comment (miivanov) [reply test/classifier_data_set_builder_test.py]: +1 keep testing logic simple

File: betel/data_set_builder.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        """Randomly splits a list of data into 3
+        (train-validation-test).
+
+        :param data: the list to be split
+        :return: the train-validation-test split in Dict format
+        """
+        app_list = self._sort(data)
+        app_list = app_list.sample(frac=1, random_state=self._RANDOM_SEED)
+
+        index_1 = int(0.7 * len(app_list))
```
Comment (tpetmanson) [inline betel/data_set_builder.py]: One more suggestion, it might be useful to make the train-test-validation split proportions configurable.

File: test/classifier_data_set_builder_test.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        {"file": storage_dir / "info" / "example", "info": "com.example,train"},
+        {"file": storage_dir / "info" / "page", "info": "com.page,train"},
+        {"file": storage_dir / "info" / "store", "info": "com.store,train"},
+        {"file": storage_dir / "info" / "play", "info": "com.play,validation"},
+        {"file": storage_dir / "info" / "example", "info": "com.test,test"}
+    ]
+
+
+class TestClassifierDataSetBuilder:
+    def test_split(self, classifier_builder):
```
Comment (miivanov) [inline test/classifier_data_set_builder_test.py:74]: is this method used?

File: test/classifier_data_set_builder_test.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        sorted_split = {}
+        for data_set in split:
+            sorted_split[data_set] = split[data_set].sort_values(by=["app_id"])
+
+        assert not all(split[data_set].equals(sorted_split[data_set]) for data_set in split)
+
+    def test_expected_locations(self, classifier_builder, input_dir, expected_icons, mocker):
+        mocker.patch.object(
+            classifier_data_set_builder.ClassifierDataSetBuilder,
+            'split',
```
Comment (miivanov) [inline test/classifier_data_set_builder_test.py]: you seem to be mocking the object that contains the most logic -> i'd expect this method to be tested and not mocked

File: betel/data_set_builder.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        :param input_dir: data to be split (output of the scraper)
+        :param storage_dir: storage directory for split data sets
+        :param split_ratio: the ratio for train-validation-test data sets
+        """
+        self._storage_dir = storage_dir
+        self._storage_dir.mkdir(exist_ok=True, parents=True)
+
+        self._input_dir = input_dir
+        self._input_file = input_dir / utils.SCRAPER_INFO_FILE_NAME
```
Comment (miivanov) [inline betel/data_set_builder.py:29]: what's the input_file contents?

File: betel/data_set_builder.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+class DataSetBuilder(metaclass=abc.ABCMeta):
+    """A class for splitting data into train-validation-test sets."""
+
+    _RANDOM_SEED = 2579  # seed for random splitting
+
+    def __init__(self, input_dir: pathlib.Path, storage_dir: pathlib.Path,
+                 split_ratio: (float, float, float) = (0.7, 0.15, 0.15)):
+        """"Constructor.
+
+        :param input_dir: data to be split (output of the scraper)
```
Comment (miivanov) [inline betel/data_set_builder.py]: Please explicitly state what's expected to be in the input_dir. It should be clear to the reader what the output of the scraper is

File: betel/classifier_data_set_builder.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import shutil
+import pathlib
+import pandas as pd
+from betel import utils
+from betel import info_files_helpers
+from betel import data_set_builder
+
+
+class ClassifierDataSetBuilder(data_set_builder.DataSetBuilder):
```
Comment (miivanov) [inline betel/classifier_data_set_builder.py:9]: could you remind me why do we need a base class here?

File: betel/data_set_builder.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        split = {
+            "train": train,
+            "validation": validation,
+            "test": test
+        }
+
+        return split
+
+    def split_and_build_data_sets(self) -> None:
+        """Splits and builds the train-validation-test data sets."""
```
Comment (miivanov) [inline betel/data_set_builder.py]: What does build mean in this case? From what I see this method outputs the input data into three different directory structures: test/train/validation -> please clearly outline this

**Review by miivanov — COMMENTED** [review]
First iteration of comments:)

One high level question: Do we really need to move images around? Is it not enough to simply split the apps into different sets and tell Keras how to load an image given an app id? It sounds a bit wasteful to copy images around

File: test/classifier_data_set_builder_test.py (hunk 7 above)
Comment (miivanov) [reply test/classifier_data_set_builder_test.py:74]: funny momo:) it's a test

File: betel/classifier_data_set_builder.py (hunk 11 above)
Comment (ingridel) [reply betel/classifier_data_set_builder.py:9]: I was thinking of another DataSetBuilder subclass for the image similarity model.

File: test/classifier_data_set_builder_test.py (hunk 8 above)
Comment (ingridel) [reply test/classifier_data_set_builder_test.py]: It is tested with the test_split method. I removed mocking from the other functions, because of the new way of testing the split function we discussed.
